File: Task/Sorting-algorithms-Shell-sort/Nim/sorting-algorithms-shell-sort.nim
proc shellSort[T](a: var openarray[T]) =
  var h = a.len
  while h > 0:
    h = h div 2
    for i in h .. < a.len:
      let k = a[i]
      var j = i
      while j >= h and k < a[j-h]:
        a[j] = a[j-h]
        j -= h
      a[j] = k

var a = @[4, 65, 2, -31, 0, 99, 2, 83, 782]
shellSort a
echo a

File: Task/Sorting-algorithms-Shell-sort/Object-Pascal/sorting-algorithms-shell-sort.pas
procedure ShellSort(var a: array of extended);
  { Sorts a vector of arbitrary length }
  { Requirement: Support for open arrays by Object Pascal compiler }
  { otherwise please use the algorithm for Pascal, which is less flexible, }
  { but also supported by Object Pascal }
var
  i, j, h, n: integer;
  v: extended;
begin
  n := length(a);
  h := 1;
  repeat
    h := 3 * h + 1
  until h > n;
  repeat
    h := h div 3;
    for i := h to n - 1 do
    begin
      v := a[i];
      j := i;
      while (j >= h) and (a[j - h] > v) do
      begin
        a[j] := a[j - h];
        j := j - h;
      end;
      a[j] := v;
    end
  until h = 1;
end;
